### PR TITLE
Add flag to prohibit equality checks between non-overlapping checks

### PR DESCRIPTION
**Pull request by ilevkivskyi**
Fixes https://github.com/python/mypy/issues/1271

The new per-file flag `--strict-equality` disables equality, identity, and container checks between non-overlapping types. In general the implementation is straightforward. Here are some corner cases I have found:
* `Any` should be always safe.
* Type promotions should be ignored, `b'abc == 'abc'` should be an error.
* Checks like `if x is not None: ...`, are special cased to not be errors if `x` has non-optional type.
* `Optional[str]` and `Optional[bytes]` should be considered non-overlapping for the purpose of this flag.
* For structural containers and custom `__eq__()` (i.e. incompatible with `object.__eq__()`) I suppress the non-overlapping types error, if there is already an error originating from the method call check.

Note that I updated `typing-full.pyi` so that `Sequence` inherits from `Container` (this is needed by some added tests, and also matches real stubs). This however caused necessary changes in a bunch of builtins fixtures.

### CODE REVIEW

**Comment by msullivan** [comment]
> Interestingly, this flag cause couple dozen errors in mypy itself, vast majority or errors are because of
> 
> ```python
> obj: Instance
> if obj is not None:
>     ...
> ```
> 

In some sense that is a good thing, right, since we are doing a bad thing in those places? On the other hand, there are a lot of them, so...

File: test-data/unit/check-expressions.test [hunk 1]
```diff
@@ -604,7 +604,7 @@ class X:
 class Y:
     def __lt__(self, o: 'Y') -> A: pass
     def __gt__(self, o: 'Y') -> A: pass
-    def __eq__(self, o: 'Y') -> B: pass
+    def __eq__(self, o: 'Y') -> B: pass  # type: ignore
```
Comment (msullivan) [inline test-data/unit/check-expressions.test]: Why is this needed?

File: mypy/checkexpr.py [hunk 2]
```diff
@@ -1954,6 +1972,26 @@ def visit_comparison_expr(self, e: ComparisonExpr) -> Type:
...
+
+        Rules:
+            * X and None are non-overlapping in strict-optional mode, and
+            overlapping otherwise.
+            * Optional[X] and Optional[Y] are non-overlapping if X and Y are
+            non-overlapping, although technically None is overlap, it is most
+            likely an error.
+            * Any overlaps with everything, i.e. always safe.
+            * Promotions are ignored, so both 'abc' == b'abc' and 1 == 1.0
+            are errors.
```
[the remark this answers is not in the document]
Comment (ilevkivskyi) [reply mypy/checkexpr.py]: As discussed off-tracker we will keep ignoring promotions, mostly for bytes vs unicode.

File: mypy/checkexpr.py [hunk 3]
```diff
@@ -1954,6 +1972,26 @@ def visit_comparison_expr(self, e: ComparisonExpr) -> Type:
...
+        """
+        if isinstance(left, UnionType) and isinstance(right, UnionType):
```
[the remark this answers is not in the document]
Comment (ilevkivskyi) [reply mypy/checkexpr.py]: Fixed.

File: test-data/unit/check-expressions.test (hunk 1 above)
Comment (ilevkivskyi) [reply test-data/unit/check-expressions.test]: I updated the fixture (to be more like in typeshed) and this now causes incompatible with supertype.

**Comment by ilevkivskyi** [comment]
> This disallows things like `assert foo is not None`

OK, also as discussed of tracker, let's start with allowing this asserts (otherwise even mypy codebase wouldn't be clean with this flag).

**Comment by ilevkivskyi** [comment]
@msullivan
Thanks for review! This is ready for another round.

File: mypy/meet.py [hunk 4]
```diff
@@ -221,9 +221,32 @@ def is_none_typevar_overlap(t1: Type, t2: Type) -> bool:
...
 
     if isinstance(left, TypeType) and isinstance(right, TypeType):
-        # TODO: Can Callable[[...], T] and Type[T] be partially overlapping?
         return _is_overlapping_types(left.item, right.item)
 
+    # Type[C] vs Callable[..., C], where the latter is class object.
+    if isinstance(left, CallableType) and left.is_type_obj() and isinstance(right, TypeType):
+        if _is_overlapping_types(left.ret_type, right.item):
+            return True
+    if isinstance(right, CallableType) and right.is_type_obj() and isinstance(left, TypeType):
```
Comment (msullivan) [inline mypy/meet.py]: I'd say define a nested function for doing the `Type[C]` vs. whatever checks instead of duplicating the logic for the left and right cases.

**Comment by JukkaL** [comment]
@ilevkivskyi I think that `int`/`float` comparison should be allowed, since they are safe and disallowing them may make this feature much less useful for code that uses lots of floats. Can you create a follow-up issue about this? It may be fine to reject `str` vs. `unicode` comparisons, but even there it's a bit questionable, since `str` and `unicode` can be freely mixed in other contexts due to the promotion. Why do we disallow this only for this particular operation?

**Comment by ilevkivskyi** [comment]
> It may be fine to reject `str` vs. `unicode` comparisons, but even there it's a bit questionable, since `str` and `unicode` can be freely mixed in other contexts due to the promotion. Why do we disallow this only for this particular operation?

This request came many times from internal users, since such comparisons caused actual troubles. One of the problems is that `b'abc' == 'abc'` is `True` in Python 2, but `False` in Python 3.

> I think that `int`/`float` comparison should be allowed, since they are safe and disallowing them may make this feature much less useful for code that uses lots of floats.

As someone who used floats a lot, I think comparing floats with `==` is a bad idea, because of traps like `(1/3)**10 * 3**10 == 1` or `(2**0.1)**10 == 2`, etc. They should be compared with something like `math.isclose()`, `numpy.allclose()` or similar.

As a bottom line, I still think prohibiting promotions with this flag is a good idea. Potentially, we can add a note why this is unsafe.

**Comment by JukkaL** [comment]
Thanks for the detailed explanation -- you convinced me. This is a great feature as it is now and will help catch many bugs.
